chrome/content/zotero/documentManager.mjs

```diff
@@ -0,0 +1,147 @@
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+    
+    Zotero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+    
+    You should have received a copy of the GNU Affero General Public License
+    along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+    
+    ***** END LICENSE BLOCK *****
+*/
+
+var { Zotero } = ChromeUtils.importESModule("chrome://zotero/content/zotero.mjs");
+
+export class DocumentManager {
+	_onSave;
+	
+	_currentState;
+	
+	_cleanState;
+	
+	_metadata;
+	
+	_unloaded = false;
+	
+	constructor(onSave) {
+		this._onSave = onSave;
+	}
+	
+	attach(win) {
+		if (win.document.getElementById('cmd_close')) {
+			throw new Error('#cmd_close already exists in document');
+		}
+		
+		let commandset = win.document.createXULElement('commandset');
+		win.document.documentElement.append(commandset);
+
+		let closeCommand = win.document.createXULElement('command');
+		closeCommand.id = 'cmd_close';
+		closeCommand.addEventListener('command', async () => {
+			if (await this.confirmClose()) {
+				win.close();
+			}
+		});
+		commandset.append(closeCommand);
+		
+		this._observer = {
+			observe: async (cancelQuit) => {
+				// We can't await a promise before deciding whether to cancel the quit,
+				// so just cancel it
+				cancelQuit.data = 1;
+				
+				// Run through the save prompt routine, and if the user doesn't choose Cancel,
+				// quit for real
+				if (await this.confirmClose()) {
+					Zotero.Utilities.Internal.quit();
+				}
+			},
+			
+			QueryInterface: ChromeUtils.generateQI([
+				"nsIObserver",
+				"nsISupportsWeakReference",
+			]),
+		};
+		Services.obs.addObserver(this._observer, 'quit-application-requested', true);
+
+		win.addEventListener('close', (event) => {
+			event.preventDefault();
+			closeCommand.doCommand();
+		});
+
+		win.addEventListener('unload', () => {
+			this._observer = null;
+			this._unloaded = true;
+		});
+	}
+
+	/**
+	 * @param {any} state The current state.
+	 * @param {Object} metadata
+	 * @param {string} [metadata.title] The document title.
+	 */
+	setState(state, metadata) {
+		this._currentState = state;
+		this._metadata = metadata;
+	}
+
+	/**
+	 * Set the current state as the "clean" state. Call this after opening/saving.
+	 */
+	setClean() {
+		this._cleanState = this._currentState;
+	}
+	
+	async confirmClose() {
+		if (this._unloaded) {
+			return true;
+		}
+		if (this._currentState === this._cleanState) {
+			return true;
+		}
+
+		let title = this._metadata.title
+			? await Zotero.ftl.formatValue('document-manager-confirm-dialog-title', {
+				title: this._metadata.title
+			})
+			: await Zotero.ftl.formatValue('document-manager-confirm-dialog-title-new');
```

Comment:
This doesn't work on Windows, where the dialog title goes in the title bar:

![Screenshot 2025-02-13 at 1 51 56 AM](https://github.com/user-attachments/assets/b5f3412f-4ce4-44ef-a1d0-0b23ec673df4)

(Not sure about Linux.)

We can use this on macOS and use something else (just the name of the editor?) for the title on Windows(/Linux) with the title string in the description, and then either add the second line after two newlines or just leave it out, since it's just there to copy macOS.